File: src/org/microg/gms/checkin/LastCheckinInfo.java
/*
 * Copyright 2013-2015 µg Project Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.microg.gms.checkin;

import android.content.Context;
import android.content.SharedPreferences;

public class LastCheckinInfo {
    public static final String PREFERENCES_NAME = "checkin";
    public static final String PREF_ANDROID_ID = "androidId";
    public static final String PREF_DIGEST = "digest";
    public static final String PREF_LAST_CHECKIN = "lastCheckin";
    public static final String PREF_SECURITY_TOKEN = "securityToken";
    public static final String INITIAL_DIGEST = "1-da39a3ee5e6b4b0d3255bfef95601890afd80709";
    public long lastCheckin;
    public long androidId;
    public long securityToken;
    public String digest;

    public static LastCheckinInfo read(Context context) {
        LastCheckinInfo info = new LastCheckinInfo();
        SharedPreferences preferences = context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE);
        info.androidId = preferences.getLong(PREF_ANDROID_ID, 0);
        info.digest = preferences.getString(PREF_DIGEST, INITIAL_DIGEST);
        info.lastCheckin = preferences.getLong(PREF_LAST_CHECKIN, 0);
        info.securityToken = preferences.getLong(PREF_SECURITY_TOKEN, 0);
        return info;
    }

    public void write(Context context) {
        context.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE).edit()
                .putLong(PREF_ANDROID_ID, androidId)
                .putString(PREF_DIGEST, digest)
                .putLong(PREF_LAST_CHECKIN, lastCheckin)
                .putLong(PREF_SECURITY_TOKEN, securityToken)
                .apply();
    }
}

File: play-services-base/core/src/main/kotlin/org/microg/gms/checkin/LastCheckinInfo.kt
/*
 * SPDX-FileCopyrightText: 2025 e foundation
 * SPDX-License-Identifier: Apache-2.0
 */

package org.microg.gms.checkin

import android.content.Context
import org.microg.gms.settings.SettingsContract

data class LastCheckinInfo(
    val lastCheckin: Long,
    val androidId: Long,
    val securityToken: Long,
    val digest: String,
    val versionInfo: String,
    val deviceDataVersionInfo: String,
) {

    constructor(r: CheckinResponse) : this(
        lastCheckin = r.timeMs ?: 0L,
        androidId = r.androidId ?: 0L,
        securityToken = r.securityToken ?: 0L,
        digest = r.digest ?: SettingsContract.CheckIn.INITIAL_DIGEST,
        versionInfo = r.versionInfo ?: "",
        deviceDataVersionInfo = r.deviceDataVersionInfo ?: "",
    )

    companion object {
        @JvmStatic
        fun read(context: Context): LastCheckinInfo {
            val projection = arrayOf(
                SettingsContract.CheckIn.ANDROID_ID,
                SettingsContract.CheckIn.DIGEST,
                SettingsContract.CheckIn.LAST_CHECK_IN,
                SettingsContract.CheckIn.SECURITY_TOKEN,
                SettingsContract.CheckIn.VERSION_INFO,
                SettingsContract.CheckIn.DEVICE_DATA_VERSION_INFO,
            )
            return SettingsContract.getSettings(
                context,
                SettingsContract.CheckIn.getContentUri(context),
                projection
            ) { c ->
                LastCheckinInfo(
                    androidId = c.getLong(0),
                    digest = c.getString(1),
                    lastCheckin = c.getLong(2),
                    securityToken = c.getLong(3),
                    versionInfo = c.getString(4),
                    deviceDataVersionInfo = c.getString(5),
                )
            }
        }

        @JvmStatic
        fun clear(context: Context) =
            SettingsContract.setSettings(context, SettingsContract.CheckIn.getContentUri(context)) {
                put(SettingsContract.CheckIn.ANDROID_ID, 0L)
                put(SettingsContract.CheckIn.DIGEST, SettingsContract.CheckIn.INITIAL_DIGEST)
                put(SettingsContract.CheckIn.LAST_CHECK_IN, 0L)
                put(SettingsContract.CheckIn.SECURITY_TOKEN, 0L)
                put(SettingsContract.CheckIn.VERSION_INFO, "")
                put(SettingsContract.CheckIn.DEVICE_DATA_VERSION_INFO, "")
            }
    }

    fun write(context: Context) =
        SettingsContract.setSettings(context, SettingsContract.CheckIn.getContentUri(context)) {
            put(SettingsContract.CheckIn.ANDROID_ID, androidId)
            put(SettingsContract.CheckIn.DIGEST, digest)
            put(SettingsContract.CheckIn.LAST_CHECK_IN, lastCheckin)
            put(SettingsContract.CheckIn.SECURITY_TOKEN, securityToken)
            put(SettingsContract.CheckIn.VERSION_INFO, versionInfo)
            put(SettingsContract.CheckIn.DEVICE_DATA_VERSION_INFO, deviceDataVersionInfo)
        }
}